packages/hooks/src/useControlled/useControlled.ts

```diff
@@ -0,0 +1,71 @@
+import { useEffect, useMemo, useState } from 'react';
+import isUndefined from 'lodash/isUndefined';
+
+import { ControlledReturnObject } from './useControlled.types';
+import { consoleOnce } from '@leafygreen-ui/lib';
+
+/**
+ * A hook that enables a component to be both controlled or uncontrolled.
+ *
+ * Returns a {@link ControlledReturnObject}
+ */
+export const useControlled = <T extends any>(
+  controlledValue?: T,
+  onChange?: (val?: T) => void,
+  initialValue?: T,
+): ControlledReturnObject<T | undefined> => {
+  /**
+   * isControlled should only be computed once
+   */
+  // eslint-disable-next-line react-hooks/exhaustive-deps
+  const isControlled = useMemo(() => !isUndefined(controlledValue), []);
+
+  /**
+   * Keep track of the uncontrolled value state internally
+   */
+  const [uncontrolledValue, setUncontrolledValue] = useState<T | undefined>(
+    initialValue,
+  );
+
+  /**
+   * The returned value.
+   * If the component is uncontrolled, it will return the internal value.
+   * If the component is controlled, it will return the controlled value.
+   */
+  const value = useMemo(
+    () => (isControlled ? controlledValue : uncontrolledValue),
+    [isControlled, uncontrolledValue, controlledValue],
+  );
+
+  /**
+   * Updates the value of the component.
+   * If the component is uncontrolled, it will update the internal value.
+   * If the component is controlled, it will not update the controlled value.
+   *
+   * onChange callback is called if provided.
+   */
+  const updateValue = (newVal: T | undefined) => {
+    if (!isControlled) {
+      setUncontrolledValue(newVal);
+    }
+    onChange?.(newVal);
+  };
+
+  /**
+   * Log a warning if neither controlled value or initialValue is provided
+   */
+  useEffect(() => {
+    if (isUndefined(controlledValue) && isUndefined(initialValue)) {
+      consoleOnce.error(
+        `Warning: \`useControlled\` hook is being used without a value or initialValue. If using an input, this will cause a React warning when an input changes. Please decide between using a controlled or uncontrolled input element, and provide either a controlledValue or initialValue to \`useControlled\``,
```

Comment:
React only shows a warning when form inputs switch between controlled and uncontrolled, so I'm thinking this message should be in the input-specific hook, but then this should also apply to input-like components using this hook. And I think we should still warn if value or initialValue is not passed in.